gazelle/gazellebinary_test.go

```diff
@@ -0,0 +1,151 @@
+/* Copyright 2019 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gazelle
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/bazel-gazelle/testtools"
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+)
+
+var (
+	gazellePath = flag.String("gazelle", "", "path to gazelle binary")
```

Comment:
> Hmm, I never thought about how hard it would be to run integration tests outside of the Gazelle main package. I wonder if there could be a better way to do this in the future. Maybe fix-update.go should be pulled out into a separate package with public functions that integration tests could call?

I'm a really big fan of creating a test macro that handles all this logic. Then if you want to write a test you just make a directory tree with these files, point it to a Gazelle binary and let a rip. It's trivial to add new testcases and this code is 95% of the way to that. It's just missing the part to pass in the `testdata` directory and make the detection of the Gazelle binary a little cleaner.

> Ideally, it should be possible to run tests without shelling out to a subprocess. go test would have a hard time caching this.

I'm not sure I understand this concern. on two levels.

1.  Why does it matter if it shells out to a subprocess? Yes, subprocesses are slow, but they are fast enough (this test runs in <0.1s on my laptop), and provide very high fidelity testing. Any go-only wrapper around this is going to have to jump through some hoops to load the plugins in the same way as the Gazelle binary ultimately would. If you do it in Go code then you have a `BUILD` file that specifies your production resources and a different `.go` file that specifies your test resources which will lead to mismatch between the two.

1.  I understand wanting to be a purist around the test runners, and I mean this as a serious question, who is the target user who likes Gazelle but doesn't like Bazel?

    It seems to me the important thing is that the tests are cacheable with the tools they are predominantly run with, which would be Bazel's test runner.